source/mir/sparse/package.d

```diff
@@ -0,0 +1,614 @@
+                table[index] = value;
+            }
+        }
+        return value;
+    }
+
+    ///
+    T opIndexUnary(string op)(size_t index)
+        if (op == `++` || op == `--`)
+    {
+        mixin (`auto value = ` ~ op ~ `table[index];`);
+        static if (isScalarType!T)
+        {
+            if (value == 0)
+            {
+                table.remove(index);
+            }
+        }
+        else
+        {
+            if (value is null)
+            {
+                table.remove(index);
+            }
+        }
+        return value;
+    }
+
+    ///
+    T opIndexOpAssign(string op)(T value, size_t index)
+        if (op == `+` || op == `-`)
+    {
+        mixin (`value = table[index] ` ~ op ~ `= value;`); // this works
+        static if (isScalarType!T)
+        {
+            if (value == 0)
+            {
+                table.remove(index);
+            }
+        }
+        else
+        {
+            if (value is null)
+            {
+                table.remove(index);
+            }
+        }
+        return value;
+    }
+}
+
+/++
++/
+struct CoordinateValue(size_t N, T)
+{
+    ///
+    size_t[N] index;
+    
+    ///
+    T value;
+
+    ///
+    sizediff_t opCmp()(auto ref const typeof(this) rht) const
+    {
+        return cmpCoo(this.index, rht.index);
+    }
+}
+
+private sizediff_t cmpCoo(size_t N)(const auto ref size_t[N] a, const auto ref size_t[N] b)
+{
+    foreach(i; Iota!(0, N))
+        if(auto d = a[i] - b[i])
+            return d;
+    return 0;
+}
+
+/++
++/
+auto byCoordinateValue(S : Slice!(N, R), size_t N, R : SparseMap!T, T)(S slice)
+{
+    static struct CoordinateValues
+    {
+        static if(N > 1)
+            private sizediff_t[N-1] _strides;
+        mixin _sparse_range_methods!(N, T);
+        private typeof(Sparse!(N, T).init.ptr.range.table.byKeyValue()) _range;
+
+        auto front() @property
+        {
+            assert(!_range.empty);
+            auto iv = _range.front;
+            size_t index = iv.key;
+            CoordinateValue!(N, T) ret = void;
+            foreach(i; Iota!(0, N - 1))
+            {
+                ret.index[i] = index / _strides[i];
+                index %= _strides[i];
+            }
+            ret.index[N - 1] = index;
+            ret.value = iv.value;
+            return ret;
+        }
+    }
+    static if(N > 1)
+    {
+        CoordinateValues ret = void;
+        ret._strides = slice.structure.strides[0..N-1];
+        ret._length = slice.ptr.range.table.length;
+        ret._range = slice.ptr.range.table.byKeyValue;
+        return ret;
+    }
+    else
+        return CoordinateValues(slice.ptr.range.table.byKeyValue);
+}
+
+///
+pure unittest
+{
+    import std.array: array;
+    import std.algorithm.sorting: sort;
+    alias CV = CoordinateValue!(2, double);
+
+    auto slice = sparse!double(3, 3);
+    slice[] = [[0, 2, 1], [0, 0, 4], [6, 7, 0]];
+    assert(slice.byCoordinateValue.array.sort().release == [
```

Comment:
maybe we should have a `flag` for `Sorted.yes` as sorting could be implemented more efficient?